src/client/datascience/kernel-launcher/localKernelFinder.ts

```diff
@@ -0,0 +1,629 @@
+} from '../jupyter/kernels/types';
+import { IJupyterKernelSpec } from '../types';
+import { ILocalKernelFinder } from './types';
+import { tryGetRealPath } from '../common';
+
+const winJupyterPath = path.join('AppData', 'Roaming', 'jupyter', 'kernels');
+const linuxJupyterPath = path.join('.local', 'share', 'jupyter', 'kernels');
+const macJupyterPath = path.join('Library', 'Jupyter', 'kernels');
+const baseKernelPath = path.join('share', 'jupyter', 'kernels');
+
+const cacheFile = 'kernelSpecPaths.json';
+
+type KernelSpecFileWithContainingInterpreter = { interpreter?: PythonEnvironment; kernelSpecFile: string };
+
+/**
+ * Helper to ensure we can differentiate between two types in union types, keeping typing information.
+ * (basically avoiding the need to case using `as`).
+ * We cannot use `xx in` as jupyter uses `JSONObject` which is too broad and captures anything and everything.
+ *
+ * @param {(nbformat.IKernelspecMetadata | PythonEnvironment)} item
+ * @returns {item is PythonEnvironment}
+ */
+export function isInterpreter(item: nbformat.INotebookMetadata | PythonEnvironment): item is PythonEnvironment {
+    // Interpreters will not have a `display_name` property, but have `path` and `type` properties.
+    return !!(item as PythonEnvironment).path && !(item as nbformat.INotebookMetadata).kernelspec?.display_name;
+}
+
+// This class searches for a kernel that matches the given kernel name.
+// First it searches on a global persistent state, then on the installed python interpreters,
+// and finally on the default locations that jupyter installs kernels on.
+@injectable()
+export class LocalKernelFinder implements ILocalKernelFinder {
+    private cache?: KernelSpecFileWithContainingInterpreter[];
+    private cacheDirty = false;
+
+    // Store our results when listing all possible kernelspecs for a resource
+    private workspaceToMetadata = new Map<string, Promise<LocalKernelConnectionMetadata[]>>();
+
+    // Store any json file that we have loaded from disk before
+    private pathToKernelSpec = new Map<string, Promise<IJupyterKernelSpec | undefined>>();
+
+    constructor(
+        @inject(IInterpreterService) private interpreterService: IInterpreterService,
+        @inject(IPlatformService) private platformService: IPlatformService,
+        @inject(IFileSystem) private fs: IFileSystem,
+        @inject(IPathUtils) private readonly pathUtils: IPathUtils,
+        @inject(IExtensionContext) private readonly context: IExtensionContext,
+        @inject(IWorkspaceService) private readonly workspaceService: IWorkspaceService,
+        @inject(IEnvironmentVariablesProvider) private readonly envVarsProvider: IEnvironmentVariablesProvider,
+        @inject(IPythonExtensionChecker) private readonly extensionChecker: IPythonExtensionChecker
+    ) {}
+    @traceDecorators.verbose('Find kernel spec')
+    @captureTelemetry(Telemetry.KernelFinderPerf)
+    public async findKernel(
+        resource: Resource,
```

Comment:
looks like this should be optional

Reply:
I'd leave this as is. `undefined` is supported value. 

This way developer needs to make a decision and explicitly pass the value in, instead of accidentally forgetting it.
Fyi, we have had many issues where these (uri) haven't been passed on when they are optional.